### DISCUSSION TITLE
[help] Strange printing behaviour running a dynamic target

### DISCUSSION THREAD

**Discussion by kendonB** [ropensci/targets#1141] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

No reprex yet as I don't have time to whittle it down. 

I run a large pipeline that ends up running a dynamic target and I see this after tar_make()

```
✔ skip target printedskippedtarget
```

Then it starts running in the background, then I interrupt using `Ctrl+C` and it prints the remaining info after the interrupt:

```
✔ skip target otherskippedtarget
✔ skip target otherskippedtarget
✔ skip target otherskippedtarget
▶ start branch startedtarget_bee8d6df
```
Any advice to make the troubleshooting/bug hunting simpler?

**Comment by kendonB** [comment]
To be clear, by 'it starts running', I mean the memory and CPU usage is consistent with it actually running the dynamic target. Furthermore, I see this a few minutes after starting the pipeline:

```
> targets::tar_progress_summary()
# A tibble: 1 × 6
  skipped started built errored canceled since
    <int>   <int> <int>   <int>    <int> <chr>
1       0       0     0       0        0 5.199 minutes
```
I wasn't seeing this behaviour when running my last project a couple of weeks ago and I did update to the latest github versions of crew, crew.cluster, and targets.

**Comment by kendonB** [comment]
Problem exists in https://github.com/ropensci/targets/commit/12948f1933417e31989cb8b4d437d30c6921ccdc but not in https://github.com/ropensci/targets/commit/3aa22618f6ed7306d62dd0db534974024b02ce3c.

**Reply by kendonB** [reply]
hopefully this helps narrow it down!

**Reply by kendonB** [reply]
@wlandau I think I've ID'd the commit in which the problem turns up

**Comment by wlandau** [comment]
I am almost positive this is part of a feature, not a bug. In the upcoming release of `targets`, `tar_make()` appends to the R console and the metadata files in batches. That means it accumulates a backlog of messages, then prints them out all at once at fixed time intervals. This behavior is a significant reduction in overhead, and it is much kinder to shared/slow file systems which are sensitive to high IOPS. The `seconds_reporter` and `seconds_meta` arguments control it, and you can set both to 0 to disable it. In practice, the intervals may be longer than `seconds_reporter` or `seconds_meta` if a target is blocking the main process, but you can make sure time-consuming operations happen on workers using `crew` along with `deployment = "worker"`, `storage = "worker"`, and `retrieval = "worker"`.

**Reply by kendonB** [reply]
You nailed it! Fixed by setting those both to 0. I guess what's happening is it gets to running the slow target so quickly that it doesn't get the chance to print. One potential change you could do is to always print before a build action? These _should_ all be decently long anyway.

**Reply by wlandau** [reply]
> One potential change you could do is to always print before a build action? These should all be decently long anyway.

The current CRAN release does this, but it is not efficient enough.

But this sort of problem would be confusing to users. So today I made these changes:

* The batched metadata/console arguments are `seconds_meta_append`, `seconds_meta_upload`, and `seconds_reporter`.
* The default values are `seconds_meta_append = 0`, `seconds_meta_upload = 15`, and `seconds_reporter = 0`.
* Details on configuring them are at the top of the performance chapter: https://github.com/ropensci-books/targets/commit/5768fb461d9c33bb5a18eb7018e02784d1a7cfce

**Reply by wlandau** [reply]
@rpodcast, these changes roll back some of the defaults we talked about, but everything is still documented and configurable. When we figure out more internal cloud stuff and it comes time to write another internal wrapper, I will set different defaults.

**Reply by kendonB** [reply]
Any insight into why I wasn't seeing anything in tar_progress_summary?

**Reply by wlandau** [reply]
> Any insight into why I wasn't seeing anything in tar_progress_summary?

Because progress data and metadata are batched too, via seconds_meta_append

**Reply by walrossker** [reply]
@wlandau I'm seeing the defaults you listed above (i.e., `seconds_meta_append = 0`, `seconds_meta_upload = 15`, and `seconds_reporter = 0`) but still not seeing target names print until after they've finished (i.e., no "start" until "built"). The behavior persists when I set `seconds_meta_upload = 0` or manually set them all to zero in the `tar_make` call. I have no concerns about the efficiency of targets running the pipeline; I just want to know which target is running at a given time.

This began after updating to these simultaneously: `targets 1.3.0`, `R 4.3.1`, and `RStudio 2023.06.2 Build 561`

**Reply by wlandau** [reply]
If you have a reprex, I can take a look. `seconds_reporter` controls the buffering of console messages.

**Reply by walrossker** [reply]
@wlandau Below is a screenshot of a currently running pipeline of my actual work with `seconds_reporter = 0`  where it's not printing which target has started, but it's clearly running something. 
![image](https://github.com/ropensci/targets/assets/35813202/4c485cf0-454a-4fa5-9aa5-93a1aef89b9c)
Strangely, when I run the toy example below with `tar_make(example_2, seconds_reporter = 0)`, it behaves normally and reports the starts and ends of both targets as they occur. Maybe it needs to be a sufficiently large or computationally intensive target to block printing?

``` r
  tar_target(example_1, {
    Sys.sleep(20)
    rnorm(1e6)
  }),
  tar_target(example_2, {
    Sys.sleep(20)
    rnorm(1e6) + example_1
  })
```

**Reply by walrossker** [reply]
Hi @wlandau, any thoughts on what might be causing print delays here? The vast majority of the time, `tar_make()` still just shows me the last target that was skipped or built

**Reply by wlandau** [reply]
I ran the example you posted above:

```r
# _targets.R:
library(targets)

list(
  tar_target(
    example_1, {
      Sys.sleep(20)
      rnorm(1e6)
    }
  ),
  tar_target(
    example_2, {
      Sys.sleep(20)
      rnorm(1e6) + example_1
    }
  )
)
``` 

And I see the right printed messages at the right times:

```r
> tar_make()
▶ start target example_1
● built target example_1 [20.048 seconds]
▶ start target example_2
● built target example_2 [20.066 seconds]
▶ end pipeline [40.669 seconds]
```

So I can't say what's happening on your end. Some speculation:

* Do you have a non-default reporter set in `_targets.yaml`? You can make sure you have the right reporter using `tar_make(reporter = "verbose")`.
* Are you running an old commit of `targets`? Maybe try the latest CRAN and latest GitHub versions.

**Reply by wlandau** [reply]
c.f. https://books.ropensci.org/targets/help.html#write-a-reprex
